### PR TITLE
RFC: Nix flake based build

### PR DESCRIPTION
**Pull request by danobi**
This is an RFC for the nix build infra we discussed during office hours.

Attached in this PR is documentation on how to interact with the nix build.
Please see that for reference.

If we decide to move forward with this, we can at the very least replace/remove:

* All the docker based build stuff in the CI
* `build-libs.sh`
* `STATIC_BPF_BCC` in cmake config
* The vendored libbpf + bcc repos

We also get for free:

* Reproducible builds / tests (modulo kernel until `vmtest` comes) between CI
  and your development host. Ideally it'll just be `nix develop --command run-tests.sh`
  that's both run in CI and locally.
* The ability to patch libbpf, bcc, llvm, really any of our deps, completely through
  `flake.nix` configs. This is something that cannot be done through traditional,
  non-functional package managers. You can sort of see in the current `flake.nix`
  how I'm patching nixpkg's upstream libbpf and bcc. This is easily extended to
  include `.patch` as well.

Downsides:

* Nix's DSL is quite difficult to understand. You can probably already see that.
  * That being said, I got somewhat productive with it in one weekend. I'd also
    be happy to walk people through what I've learned.
  * It's also quite beautiful (the semantics, not the syntax) when you piece it all together.
* One more thing to learn (but hopefully not maintain, as we'll delete stuff if this goes in)
  * But the industry is moving towards this direction so hopefully another tool in your tool kit

What's left to do:

* Fix all the tests
* Hook up to CI
* Delete all the will-be-deprecated stuff

##### Checklist

- [ ] Language changes are updated in `man/adoc/bpftrace.adoc` and if needed in `docs/reference_guide.md`
- [ ] User-visible and non-trivial changes updated in `CHANGELOG.md`
- [ ] The new behaviour is covered by tests

### CODE REVIEW

File: README.md
Comment (JakeHillion) [inline README.md:164]: I'm not sure it's great to include the install command here, particularly as it ends up using high privilege at points. Probably best to link out to the Nix download page as you have above.

File: flake.nix
Comment (JakeHillion) [inline flake.nix:53]: Do we need these overlays? It appears to at least build fine with the current versions in the `nixpkgs` referred to by the `flake.lock` file. I agree that they may be needed at some point in the future so perhaps it is better to have the infrastructure, and I understand the motivation to keep them synced with the submodules. Worth a discussion at least.

File: flake.nix
Comment (JakeHillion) [inline flake.nix:165]: If these stay, `s/x86_64-linux/"${system}"/`

File: flake.nix
Comment (JakeHillion) [inline flake.nix:10]: I think it's reasonable to limit this to tested systems in the beginning, but I don't think this is true of bpftrace - it supports quite a few architectures.

**Review by JakeHillion — COMMENTED** [review]
Thank you for doing this, more Nix flake adoption in upstream repos is great to see! I commented on some more specific parts but generally.
- Rather than including the tests in the binary output they should really be included as flake checks (`outputs.checks."${system}".tests = ...`). This will make the whole local testing/CI run with `nix flake check`.
- I'm not entirely convinced by the default dev build mode. My assumption would be that the build is a prod build, with a user dropping into a dev shell (`nix develop`) if they need anything special. Worth discussing.
- Having `bpftrace` in the `apps` part of the output would be nice, allowing `nix run`. This wouldn't interact very well with `sudo` though. Again, worth discussing.
- I don't think we need `devShells.default` at all in its current form, as it's effectively set to the default (`+runTests`, which can be covered with `nix flake check`). Not sure how other projects have handled this though.
- I think there are some dependencies missing as `cmake -B build; cd build; make` didn't work on my NixOS system in `nix develop` (a really clean external path). This is minutia though.

The output schema is [here](https://nixos.wiki/wiki/Flakes#Output_schema).

Great starting point, and thanks for offering up this RFC!

File: flake.nix
Comment (RyanGibb) [reply flake.nix:165]: `s/x86_64-linux/${system}/` should be fine (the quotes aren't necessary)

File: README.md
Comment (danobi) [reply README.md:164]: Ack. Best to let the nix download page sell the mechanism.

File: flake.nix
Comment (danobi) [reply flake.nix:53]: Yes, I think we need these. Based on the discussion in #2334 and #2322 there is consensus that bpftrace should have a strong dependency on libbpf, especially to ease development and migration away from bcc. It is too difficult to support every libbpf under the sun and perform possibly the largest refactor this project has seen. During release time we will simply bump the required version to the one we developed against (which solve the packaging problem too).

File: flake.nix
Comment (danobi) [reply flake.nix:165]: Ack. Good catch.

File: flake.nix
Comment (danobi) [reply flake.nix:10]: I think nix only supports aarch and x86 on linux and osx, right? We do support quite a few architectures but none 32 bit. And osx has no bpf support.

So that limits nix support to aarch64-linux and x86-64-linux, right? Or am I misunderstanding something?

**Comment by danobi** [comment]
Hi @JakeHillion, @RyanGibb , thank you for the reviews! It is much appreciated b/c I am quite new to nix.

> Thank you for doing this, more Nix flake adoption in upstream repos is great to see! I commented on some more specific parts but generally.
> 
>     * Rather than including the tests in the binary output they should really be included as flake checks (`outputs.checks."${system}".tests = ...`). This will make the whole local testing/CI run with `nix flake check`.

Thanks for the tip. I will look into that. Looks like it's uncovering a few (maybe bad) assumptions already b/c it's running in a hermetic environment.

>     * I'm not entirely convinced by the default dev build mode. My assumption would be that the build is a prod build, with a user dropping into a dev shell (`nix develop`) if they need anything special. Worth discussing.

So the goal of this flake is to address some the pain points in CI we have. I'd like to start off with fixing development pain first and then if possible, make it a flake suitable for downstream users. Unless there's an easy way to satisfy both that I'm missing (and I probably am), I'd prefer to stick with dev in mind first.

>     * Having `bpftrace` in the `apps` part of the output would be nice, allowing `nix run`. This wouldn't interact very well with `sudo` though. Again, worth discussing.

Ack. Might as well add it even if really only --help works.

>     * I don't think we need `devShells.default` at all in its current form, as it's effectively set to the default (`+runTests`, which can be covered with `nix flake check`). Not sure how other projects have handled this though.

Yeah makes sense. Will address it when I fix `nix flake check`.

>     * I think there are some dependencies missing as `cmake -B build; cd build; make` didn't work on my NixOS system in `nix develop` (a really clean external path). This is minutia though.

Hmm, that's weird. It's the same deps as `nix build` right? And that seems to work ok.

[...]

**Comment by danobi** [comment]
This patch makes it so tests are run (but they fail for various reasons): https://pastes.dxuuu.xyz/f8tnu6

Running into this issue: 
```
$ nix flake check --allow-new-privileges
warning: Git tree '/home/dxu/dev/bpftrace' is dirty
error: builder for '/nix/store/b2px0hqb5i8l3m52jii5za99rg7bmy0z-bpftrace.drv' failed with exit code 1;
       last 10 log lines:
       > [192/196] Linking CXX static library src/ast/libast.a
       > [193/196] Linking CXX static library src/libbpftrace.a
       > [194/196] Linking CXX executable src/bpftrace
       > /nix/store/9n9gjvzci75gp2sh1c4rh626dhizqynl-binutils-2.39/bin/ld: warning: libbpf.so.0, needed by /nix/store/qbrgk50h4z64vxfdx5z7a31knwszsd9n-bcc-0.27.0/lib/libbcc_bpf.so, may conflict with libbpf.so.1
       > [195/196] Building CXX object tests/CMakeFiles/bpftrace_test.dir/semantic_analyser.cpp.o
       > [196/196] Linking CXX executable tests/bpftrace_test
       > /nix/store/9n9gjvzci75gp2sh1c4rh626dhizqynl-binutils-2.39/bin/ld: warning: libbpf.so.0, needed by /nix/store/qbrgk50h4z64vxfdx5z7a31knwszsd9n-bcc-0.27.0/lib/libbcc_bpf.so, may conflict with libbpf.so.1
       > running tests
       > sudo: The "no new privileges" flag is set, which prevents sudo from running as root.
       > sudo: If sudo is running in a container, you may need to adjust the container configuration to disable the flag.
       For full logs, run 'nix-store -l /nix/store/b2px0hqb5i8l3m52jii5za99rg7bmy0z-bpftrace.drv'.
```

Can't seem to get sudo to work even with `--allow-new-privileges`

File: flake.nix
Comment (JakeHillion) [reply flake.nix:53]: I think you would still have precise control as there is control over when the lock and hence `nixpkgs` dependency is updated, but this definitely makes it more explicit. Fine by me.

**Review by JakeHillion — APPROVED** [review]
Thanks for the great discussion! With this enabled in the CI and the tests fixed this looks good to me, and gives us a fantastic base to add new Nix features in the future.

**Comment by danobi** [comment]
> I think the best solution to this long term is to run the tests in VMs using https://nixos.org/manual/nixos/stable/index.html#sec-nixos-tests. However this has a bit of a conflict of interest with your existing https://github.com/iovisor/bpftrace/pull/2607. As far as I can tell, it would be possible to achieve the same thing by pinning kernels in the NixOS testing framework.

Looks neat. Will take a look. That looks like a great option when the entire workflow has already been onboarded to nix.

That being said, there's 1 big blocker in the way of all the VM testing stuff: nested virtualization support. Right now it took over 6 hours to run the runtime tests (and it still didn't finish) in a VM in standard GHA. GH supports large runners with nested virt now but we need a paid plan.

**Comment by danobi** [comment]
I asked around in the nix matrix room and thought about it overnight. Basically they're saying that flake checks cannot run anything that requires root. Disappointing but understandable.

In light of this, I think we should scale back the ambitions of the nix flake build a bit. I've pushed changes to make it more of a dependency manager rather than an all-in-one CI thing.  This will still immediately solve some developer pain. We'll just have multiple steps in the CI that runs stuff inside nix developer shells rather than just with `nix flake check`.

I've also incorporated the feedback about making the flake geared towards downstream users. The cmake flags and such will be manually chosen in the developer shell now.

**Comment by danobi** [comment]
@viktormalik @fbs @ajor any thoughts? At the very least this would provide an easy way to get started with developing on bpftrace.

I am planning on replacing `ci.yml` workflow using this nix flake (to get better local reproducibility). I am also hoping we can replace the embedded builds as well with the help of https://github.com/rsms/llvmbox . But I am less certain about embedded builds.

I have no strong opinion on whether we should wait for ci.yml replacement work to be done before merging this, but like I mentioned, the flake here is enough for smoother onboarding of new developers.

**Comment by JakeHillion** [comment]
Ack these changes, I think this looks a lot nicer now. Agreed on the scaled back ambitions too, as everybody gets a clean development environment with consistent dependencies and can work from there.

With luck fixing the sudo issues can be solved with some form of vm testing, but if this goes in as is it'll be an excellent base for both new developers and the CI - extensions can come later.

File: flake.nix
Comment (ajor) [reply flake.nix:10]: bpftrace does support 32-bit ARM now, FWIW

**Review by ajor — COMMENTED** [review]
I know no Nix, but I like the sound of it. I'd say to go ahead and merge this before fiddling with the CI - what we've got here is useable right now.

I think the Docker-based build system will still be useful for confirming that bpftrace builds on various Linux distros, so wouldn't get rid of it entirely. It seems fair enough to move the general CI builds and the LLVM matrix across to Nix though.

**Comment by danobi** [comment]
> I think the Docker-based build system will still be useful for confirming that bpftrace builds on various Linux distros, so wouldn't get rid of it entirely. It seems fair enough to move the general CI builds and the LLVM matrix across to Nix though.

I agree it's useful in principle. But I'd like to point out in practice that `ci.yml` only builds with ubuntu 20.04 and `embedded.yml` builds with ubuntu 18.04, 20.04, and alpine 3.12. So at minimum reworking `ci.yml` is status-quo.

Long term I'd like to revisit the embedded llvm builds. We pay a heavy maintenance cost for that. I am hoping (praying perhaps) that we can externalize that cost to llvmbox. But that will come later, if at all.

File: flake.nix
Comment (danobi) [reply flake.nix:101]: Ack.

**Pull request merged**
